src/main/java/org/deepsymmetry/beatlink/VirtualRekordbox.java

```diff
@@ -280,13 +286,98 @@ SlotReference findMatchedTrackSourceSlotForPlayer(int player) {
     }
 
+    /**
+     * Given a player number (normalized to the range 1-4), returns the DeviceSQL rekordbox ID for the
+     * track that is currently loaded on that player. Returns 0 if we are searching for a PSSI match still on a new track load,
+     * or if a PSSI match wasn't found, or if the player does not have a track loaded.
+     *
+     * @param player the player whose DeviceSQL rekordbox ID we are interested in
+     * @return the DeviceSQL rekordbox ID for the track that is currently loaded on that player
+     */
+    int findDeviceSqlRekordboxIdForPlayer(int player) {
+        return playerToDeviceSqlRekordboxId.getOrDefault(player, 0);
+    }
 
     /**
      * Keeps track of the most recent valid (non-zero) status flag byte we have received from each device number,
      * so we can reuse it in cases where a corrupt (zero) value has been sent to us.
      */
     private final Map<Integer, Byte> lastValidStatusFlagBytes = new ConcurrentHashMap<>();
 
+    /**
+     * Tracks packets received from devices to reconstruct complete messages that span multiple packets.
+     * Used primarily for reconstructing binary data (specifically PSSI) from the Opus Quad, which 
+     * arrives fragmented across multiple packets. Maintains state between packet
+     * arrivals until a complete message is assembled, at which point the data can be processed.
+     */
+    private static class PacketTracker {
+        private final List<Byte> data;
+
+        PacketTracker() {
+            this.data = new ArrayList<>();
+        }
+
+        /**
+         * Receives and processes a new packet in the sequence.
+         *
+         * @param packetData the data from the new packet
+         * @param packetNumber which packet this is in the sequence
+         * @param totalPackets how many packets make up the complete message
+         * @return true if this was the final packet in the sequence
+         */
+        boolean receivePacket(byte[] packetData, int packetNumber, int totalPackets) {
+            for (byte b : packetData) {
+                data.add(b);
+            }
+
+            if (packetNumber == totalPackets) {
+                return true;
+            }
+            return false;
+        }
+
+        /**
+         * Reset the tracker state
+         */
+        void resetForNewPacketStream() {
+            data.clear();
+        }
+
+        /**
+         * Convert the accumulated data to a byte array, trimming trailing zeros.
+         * We are not sure if the OPUS_METADATA packets send the length of the entire PSSI back,
+         * so for now we just trim the trailing zeros.
+         *
+         * @return the complete message data with trailing zeros removed
+         */
+        byte[] getDataAsBytesAndTrimTrailingZeros() {
+            // First convert to regular byte array
+            byte[] result = new byte[data.size()];
+            for (int i = 0; i < data.size(); i++) {
```

Comment:
Again, instead of writing your own loop here, you can just say:
`byte[] result = data.toArray();`

And this isn’t necessary, but you might want to remove the trailing zeroes while it is still an `ArrayList` by using a while loop to `remove()` the last element as long as the last element is a zero, and then just call `toArray()` after doing that. It would be another array copy operation we could eliminate.

Reply:
For similar reasons to the above, since your array list wraps the bytes as Byte objects, we can’t just simply make it an array any better than you did, but this fact is increasing my motivation to switch this stuff to use a `ByteBuffer`. However, I will resist doing that until we have merged your PR and someone says they are willing to test my “improvements”. I did swap the order of trimming the trailing zeroes, and that makes things a bit cleaner and takes better advantage of the array list, eliminating one array copy. I’ll push that, link the commit here, and resolve this thread.

Reply:
All right, here is the tweak I made (it also included a small optimization to `receivePacket`: https://github.com/Deep-Symmetry/beat-link/pull/86/commits/b7e1455e7874eb32a601c110b48fc22be9fb8218